src/services/DealService.ts

```diff
@@ -108,7 +110,52 @@ export class DealService {
     /**
      * deals will take care of themselves on account changes
      */
-    this.getDeals();
+    if (!this.deals) {
+      await this.getDeals();
+    }
+
+    this.observeDeals();
+  }
+
+  private async observeDeals(): Promise<void> {
+    this.deals = new Map<Address, DealTokenSwap>();
```

Comment:
You should not set this.deals until the Map is completely populated.  Use a local Map and assign at the end, like getDeals does.  This will help minimize thrashing.